From 90a8f9326211e2ad52864afe60b8f2d02c024169 Mon Sep 17 00:00:00 2001
From: Peter Zotov <whitequark@whitequark.org>
Date: Fri, 17 Oct 2014 08:05:26 +0400
Subject: [PATCH 2/4] Detect host ocamlyacc alongside host ocamlrun in
 configure.

---
 configure | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/configure b/configure
index a3909b6..389430c 100755
--- a/configure
+++ b/configure
@@ -467,14 +467,17 @@ if test x"$target" != x"$host"; then
           "($ocaml_system_version) doesn't match the version of these sources\n" \
           "($ocaml_source_version)."
     else
+      CAMLYACC="ocamlyacc"
       CAMLRUN="ocamlrun"
     fi
   fi
 else
   CAMLRUN=`cd ../.. && pwd`/boot/ocamlrun
+  CAMLYACC=`cd ../.. && pwd`/boot/ocamlyacc
 fi
 
 echo "CAMLRUN=$CAMLRUN" >> Makefile
+echo "CAMLYACC=$CAMLYACC" >> Makefile
 
 # Check the sizes of data types
 # OCaml needs a 32 or 64 bit architecture, a 32-bit integer type and
-- 
2.1.1

