--- OCamlMakefile.orig	2013-08-16 11:18:04.000000000 +0000
+++ OCamlMakefile	2016-12-18 18:45:38.562159381 +0000
@@ -297,10 +297,11 @@
       OCAML_DEP_PACKAGES :=
     endif
     OCAML_FIND_LINKPKG := -linkpkg
-    REAL_OCAMLFIND := $(OCAMLFIND)
   endif
 endif
 
+REAL_OCAMLFIND := $(OCAMLFIND)
+
 export OCAML_FIND_PACKAGES
 export OCAML_DEP_PACKAGES
 export OCAML_FIND_LINKPKG
@@ -885,7 +886,7 @@
 
 else
 $(DLLSONAME):		$(OBJ_LINK)
-			$(OCAMLMKLIB) $(INCFLAGS) $(CLIBFLAGS) \
+			$(REAL_OCAMLFIND) $(OCAMLMKLIB) $(INCFLAGS) $(CLIBFLAGS) \
 				-o $(CLIB_BASE) $(OBJ_LINK) $(CLIBS:%=-l%) \
 				$(OCAMLMKLIB_FLAGS)
 endif
@@ -1010,9 +1011,9 @@
 			$(QUIET)if [ $(NOIDLHEADER) ]; then touch $*.h; fi
 
 .c.$(EXT_OBJ):
-			$(OCAMLC) -c -cc "$(CC)" -ccopt "$(CFLAGS) \
+			$(REAL_OCAMLFIND) $(OCAMLC) -c -ccopt "$(CFLAGS) \
 				$(CPPFLAGS) $(CPPFLAGS_WIN32) \
-				$(CFLAGS_WIN32) $(CINCFLAGS) $(CFLAG_O)$@ " $< 
+				$(CFLAGS_WIN32) $(CINCFLAGS) $(CFLAG_O)$@ " $<
 
 .$(EXT_CXX).$(EXT_OBJ):
 			$(CXX) -c $(CXXFLAGS) $(CINCFLAGS) $(CPPFLAGS) \
