diff --git a/Makefile b/Makefile
index 1c7a07e..ae47c66 100644
--- a/Makefile
+++ b/Makefile
@@ -15,18 +15,16 @@
 include config/Makefile
 include stdlib/StdlibModules
 
-CAMLC=boot/ocamlrun boot/ocamlc -nostdlib -I boot
-CAMLOPT=boot/ocamlrun ./ocamlopt -nostdlib -I stdlib -I otherlibs/dynlink
+CAMLC=$(CAMLRUN) boot/ocamlc -nostdlib -I boot -use-runtime $(CAMLRUN)
+CAMLOPT=$(CAMLRUN) ./ocamlopt -nostdlib -I stdlib -I otherlibs/dynlink
 COMPFLAGS=-strict-sequence -w +33..39+48 -warn-error A -bin-annot \
           -safe-string $(INCLUDES)
 LINKFLAGS=
 
-CAMLYACC=boot/ocamlyacc
 YACCFLAGS=-v
-CAMLLEX=boot/ocamlrun boot/ocamllex
-CAMLDEP=boot/ocamlrun tools/ocamldep
+CAMLLEX=$(CAMLRUN) boot/ocamllex
+CAMLDEP=$(CAMLRUN) tools/ocamldep
 DEPFLAGS=$(INCLUDES)
-CAMLRUN=byterun/ocamlrun
 SHELL=/bin/sh
 MKDIR=mkdir -p
 
@@ -652,7 +650,7 @@ beforedepend:: asmcomp/emit.ml
 
 tools/cvt_emit: tools/cvt_emit.mll
 	cd tools; \
-	$(MAKE) CAMLC="../$(CAMLRUN) ../boot/ocamlc -I ../stdlib" cvt_emit
+	$(MAKE) CAMLC="$(CAMLRUN) ../boot/ocamlc -use-runtime $(CAMLRUN) -I ../stdlib" cvt_emit
 
 # The "expunge" utility
 
diff --git a/Makefile.nt b/Makefile.nt
index 6fe9f60..884b639 100644
--- a/Makefile.nt
+++ b/Makefile.nt
@@ -15,16 +15,14 @@
 include config/Makefile
 include stdlib/StdlibModules
 
-CAMLC=boot/ocamlrun boot/ocamlc -nostdlib -I boot
-CAMLOPT=boot/ocamlrun ./ocamlopt -nostdlib -I stdlib -I otherlibs/dynlink
+CAMLC=$(CAMLRUN) boot/ocamlc -nostdlib -I boot
+CAMLOPT=$(CAMLRUN) ./ocamlopt -nostdlib -I stdlib -I otherlibs/dynlink
 COMPFLAGS=-strict-sequence -w +33..39+48 -warn-error A -bin-annot $(INCLUDES)
 LINKFLAGS=
-CAMLYACC=boot/ocamlyacc
 YACCFLAGS=
-CAMLLEX=boot/ocamlrun boot/ocamllex
-CAMLDEP=boot/ocamlrun tools/ocamldep
+CAMLLEX=$(CAMLRUN) boot/ocamllex
+CAMLDEP=$(CAMLRUN) tools/ocamldep
 DEPFLAGS=$(INCLUDES)
-CAMLRUN=byterun/ocamlrun
 
 OCAMLBUILDBYTE=$(WITH_OCAMLBUILD:=.byte)
 OCAMLBUILDNATIVE=$(WITH_OCAMLBUILD:=.native)
@@ -331,8 +329,9 @@ partialclean::
 	rm -f compilerlibs/ocamlbytecomp.cma
 
 ocamlc: compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma $(BYTESTART)
-	$(CAMLC) $(LINKFLAGS) -compat-32 -o ocamlc compilerlibs/ocamlcommon.cma \
-	         compilerlibs/ocamlbytecomp.cma $(BYTESTART)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -compat-32 \
+           -o ocamlc compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma \
+           $(BYTESTART)
 
 partialclean::
 	rm -f ocamlc
@@ -345,8 +344,8 @@ partialclean::
 	rm -f compilerlibs/ocamloptcomp.cma
 
 ocamlopt: compilerlibs/ocamlcommon.cma compilerlibs/ocamloptcomp.cma $(OPTSTART)
-	$(CAMLC) $(LINKFLAGS) -o ocamlopt \
-          compilerlibs/ocamlcommon.cma compilerlibs/ocamloptcomp.cma $(OPTSTART)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o ocamlopt \
+           compilerlibs/ocamlcommon.cma compilerlibs/ocamloptcomp.cma $(OPTSTART)
 
 partialclean::
 	rm -f ocamlopt
@@ -360,7 +359,7 @@ partialclean::
 
 ocaml: compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma \
        compilerlibs/ocamltoplevel.cma $(TOPLEVELSTART) expunge
-	$(CAMLC) $(LINKFLAGS) -linkall -o ocaml.tmp \
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -linkall -o ocaml.tmp \
           compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma \
           compilerlibs/ocamltoplevel.cma $(TOPLEVELSTART)
 	- $(CAMLRUN) ./expunge ocaml.tmp ocaml $(PERVASIVES)
@@ -580,7 +579,8 @@ tools/cvt_emit: tools/cvt_emit.mll
 
 expunge: compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma \
          toplevel/expunge.cmo
-	$(CAMLC) $(LINKFLAGS) -o expunge compilerlibs/ocamlcommon.cma \
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o expunge compilerlibs/ocamlcommon.cma \
 	         compilerlibs/ocamlbytecomp.cma toplevel/expunge.cmo
 
 partialclean::
diff --git a/configure b/configure
index 1be4682..4ded701 100755
--- a/configure
+++ b/configure
@@ -482,6 +482,24 @@ fi
 
 echo "CAMLRUN=$CAMLRUN" >> Makefile
 
+# Same holds for ocamlyacc.
+if test x"$target" != x"$host"; then
+  if ! sh ./searchpath ocamlyacc; then
+    err "Cross-compilation requires an ocamlyacc binary."
+  else
+    ocamlyacc 2>/dev/null
+    if test "$?" -ne 1; then
+      err "While you have an ocamlyacc binary, it cannot be executed successfully."
+    else
+      CAMLYACC="ocamlyacc"
+    fi
+  fi
+else
+  CAMLYACC=`cd ../.. && pwd`/boot/ocamlyacc
+fi
+
+echo "CAMLYACC=$CAMLYACC" >> Makefile
+
 # Check the sizes of data types
 # OCaml needs a 32 or 64 bit architecture, a 32-bit integer type and
 # a 64-bit integer type
diff --git a/debugger/Makefile.shared b/debugger/Makefile.shared
index fed1d26..9c9991a 100644
--- a/debugger/Makefile.shared
+++ b/debugger/Makefile.shared
@@ -12,14 +12,12 @@
 
 include ../config/Makefile
 
-ROOTDIR=..
-CAMLC=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
+CAMLC=$(CAMLRUN) ../ocamlc -nostdlib -I ../stdlib
 COMPFLAGS=-warn-error A -safe-string $(INCLUDES)
 LINKFLAGS=-linkall -I $(UNIXDIR)
-CAMLYACC=../boot/ocamlyacc
 YACCFLAGS=
-CAMLLEX=../boot/ocamlrun ../boot/ocamllex
-CAMLDEP=../boot/ocamlrun ../tools/ocamldep
+CAMLLEX=$(CAMLRUN) ../boot/ocamllex
+CAMLDEP=$(CAMLRUN) ../tools/ocamldep
 DEPFLAGS=$(INCLUDES)
 
 INSTALL_BINDIR=$(DESTDIR)$(BINDIR)
@@ -83,7 +81,8 @@ OBJS=\
 all: ocamldebug$(EXE)
 
 ocamldebug$(EXE): $(OBJS) $(OTHEROBJS)
-	$(CAMLC) $(LINKFLAGS) -o ocamldebug$(EXE) -linkall $(OTHEROBJS) $(OBJS)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o ocamldebug$(EXE) -linkall $(OTHEROBJS) $(OBJS)
 
 install:
 	cp ocamldebug$(EXE) $(INSTALL_BINDIR)/ocamldebug$(EXE)
diff --git a/lex/Makefile b/lex/Makefile
index cb5df8b..c1bf0c3 100644
--- a/lex/Makefile
+++ b/lex/Makefile
@@ -11,13 +11,15 @@
 #########################################################################
 
 # The lexer generator
-CAMLC=../boot/ocamlrun ../boot/ocamlc -strict-sequence -nostdlib -I ../boot
-CAMLOPT=../boot/ocamlrun ../ocamlopt -nostdlib -I ../stdlib
+include ../config/Makefile
+
+CAMLC=$(CAMLRUN) ../boot/ocamlc -strict-sequence -nostdlib -I ../boot
+CAMLOPT=$(CAMLRUN) ../ocamlopt -nostdlib -I ../stdlib
 COMPFLAGS=-w +33..39 -warn-error A -bin-annot -safe-string
-CAMLYACC=../boot/ocamlyacc
+LINKFLAGS=
 YACCFLAGS=-v
-CAMLLEX=../boot/ocamlrun ../boot/ocamllex
-CAMLDEP=../boot/ocamlrun ../tools/ocamldep
+CAMLLEX=$(CAMLRUN) ../boot/ocamllex
+CAMLDEP=$(CAMLRUN) ../tools/ocamldep
 
 
 OBJS=cset.cmo syntax.cmo parser.cmo lexer.cmo table.cmo lexgen.cmo \
@@ -27,7 +29,7 @@ all: ocamllex
 allopt: ocamllex.opt
 
 ocamllex: $(OBJS)
-	$(CAMLC) $(LINKFLAGS) -compat-32 -o ocamllex $(OBJS)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -compat-32 -o ocamllex $(OBJS)
 
 ocamllex.opt: $(OBJS:.cmo=.cmx)
 	$(CAMLOPT) -o ocamllex.opt $(OBJS:.cmo=.cmx)
diff --git a/lex/Makefile.nt b/lex/Makefile.nt
index 38c71f2..53b43c6 100644
--- a/lex/Makefile.nt
+++ b/lex/Makefile.nt
@@ -14,14 +14,13 @@
 
 include ../config/Makefile
 
-CAMLC=../boot/ocamlrun ../boot/ocamlc -I ../boot
-CAMLOPT=../boot/ocamlrun ../ocamlopt -I ../stdlib
+CAMLC=$(CAMLRUN) ../boot/ocamlc -I ../boot
+CAMLOPT=$(CAMLRUN) ../ocamlopt -I ../stdlib
 COMPFLAGS=-warn-error A
 LINKFLAGS=
-CAMLYACC=../boot/ocamlyacc
 YACCFLAGS=-v
-CAMLLEX=../boot/ocamlrun ../boot/ocamllex
-CAMLDEP=../boot/ocamlrun ../tools/ocamldep
+CAMLLEX=$(CAMLRUN) ../boot/ocamllex
+CAMLDEP=$(CAMLRUN) ../tools/ocamldep
 DEPFLAGS=
 
 OBJS=cset.cmo syntax.cmo parser.cmo lexer.cmo table.cmo lexgen.cmo \
@@ -31,7 +30,7 @@ all: ocamllex syntax.cmo
 allopt: ocamllex.opt
 
 ocamllex: $(OBJS)
-	$(CAMLC) $(LINKFLAGS) -compat-32 -o ocamllex $(OBJS)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -compat-32 -o ocamllex $(OBJS)
 
 ocamllex.opt: $(OBJS:.cmo=.cmx)
 	$(CAMLOPT) -o ocamllex.opt $(OBJS:.cmo=.cmx)
diff --git a/ocamlbuild/Makefile b/ocamlbuild/Makefile
index b40d0ea..0f596eb 100644
--- a/ocamlbuild/Makefile
+++ b/ocamlbuild/Makefile
@@ -13,11 +13,10 @@
 include ../config/Makefile
 
 ROOTDIR   = ..
-OCAMLRUN  = $(ROOTDIR)/boot/ocamlrun
-OCAMLC    = $(OCAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLOPT  = $(OCAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLDEP  = $(OCAMLRUN) $(ROOTDIR)/tools/ocamldep
-OCAMLLEX  = $(OCAMLRUN) $(ROOTDIR)/boot/ocamllex
+OCAMLC    = $(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLOPT  = $(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLDEP  = $(CAMLRUN) $(ROOTDIR)/tools/ocamldep
+OCAMLLEX  = $(CAMLRUN) $(ROOTDIR)/boot/ocamllex
 CP        = cp
 COMPFLAGS= -warn-error A -w L -w R -w Z -I ../otherlibs/$(UNIXLIB) -safe-string
 LINKFLAGS= -I ../otherlibs/$(UNIXLIB)
@@ -98,11 +97,11 @@ allopt: ocamlbuild.native ocamlbuildlib.cmxa
 # The executables
 
 ocamlbuild.byte: ocamlbuild_pack.cmo $(EXTRA_CMO) ocamlbuild.cmo
-	$(OCAMLC) $(LINKFLAGS) -o ocamlbuild.byte \
+	$(OCAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o ocamlbuild.byte \
           unix.cma ocamlbuild_pack.cmo $(EXTRA_CMO) ocamlbuild.cmo
 
 ocamlbuildlight.byte: ocamlbuild_pack.cmo ocamlbuildlight.cmo
-	$(OCAMLC) $(LINKFLAGS) -o ocamlbuildlight.byte \
+	$(OCAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o ocamlbuildlight.byte \
           ocamlbuild_pack.cmo ocamlbuildlight.cmo
 
 ocamlbuild.native: ocamlbuild_pack.cmx $(EXTRA_CMX) ocamlbuild.cmx
diff --git a/ocamlbuild/Makefile.noboot b/ocamlbuild/Makefile.noboot
deleted file mode 100644
index 313e568..0000000
--- a/ocamlbuild/Makefile.noboot
+++ /dev/null
@@ -1,227 +0,0 @@
-#(***********************************************************************)
-#(*                                                                     *)
-#(*                             ocamlbuild                              *)
-#(*                                                                     *)
-#(*                           Wojciech Meyer                            *)
-#(*                                                                     *)
-#(*  Copyright 2012 Institut National de Recherche en Informatique et   *)
-#(*  en Automatique.  All rights reserved.  This file is distributed    *)
-#(*  under the terms of the Q Public License version 1.0.               *)
-#(*                                                                     *)
-#(***********************************************************************)
-
-# This file removes the dependency on ocamlbuild itself, thus removes need
-# for bootstrap. The base for this Makefile was ocamldoc Makefile.
-
-include ../config/Makefile
-
-# Various commands and dir
-##########################
-
-ROOTDIR   = ..
-OCAMLRUN  = $(ROOTDIR)/boot/ocamlrun
-OCAMLC    = $(OCAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLOPT  = $(OCAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLDEP  = $(OCAMLRUN) $(ROOTDIR)/tools/ocamldep
-OCAMLLEX  = $(OCAMLRUN) $(ROOTDIR)/boot/ocamllex
-OCAMLLIB = $(LIBDIR)
-OCAMLBIN = $(BINDIR)
-
-# For installation
-##############
-MKDIR=mkdir -p
-CP=cp -f
-OCAMLBUILD=ocamlbuild
-OCAMLBUILD_OPT=$(OCAMLBUILD).opt
-OCAMLBUILD_LIBCMA=ocamlbuildlib.cma
-OCAMLBUILD_LIBCMI=ocamlbuildlib.cmi
-OCAMLBUILD_LIBCMXA=ocamlbuild.cmxa
-OCAMLBUILD_LIBA=ocamlbuild.$(A)
-INSTALL_LIBDIR=$(DESTDIR)$(OCAMLLIB)/ocamlbuild
-INSTALL_CUSTOMDIR=$(INSTALL_LIBDIR)/custom
-INSTALL_BINDIR=$(DESTDIR)$(OCAMLBIN)
-
-INSTALL_MLIS=
-INSTALL_CMIS=$(INSTALL_MLIS:.mli=.cmi)
-
-# Compilation
-#############
-OCAMLSRCDIR=..
-INCLUDES_DEP=
-
-INCLUDES_NODEP=	-I $(OCAMLSRCDIR)/stdlib \
-	-I $(OCAMLSRCDIR)/otherlibs/str \
-	-I $(OCAMLSRCDIR)/otherlibs/dynlink \
-	-I $(OCAMLSRCDIR)/otherlibs/$(UNIXLIB)
-
-INCLUDES=$(INCLUDES_DEP) $(INCLUDES_NODEP)
-
-COMPFLAGS=$(INCLUDES) -warn-error A -safe-string
-LINKFLAGS=$(INCLUDES)
-
-CMOFILES_PACK= 			            \
-	 ocamlbuild_Myocamlbuild_config.cmo \
-	 discard_printf.cmo 		    \
-         my_std.cmo 			    \
-         bool.cmo                           \
-         glob_ast.cmo                       \
-         glob_lexer.cmo                     \
-         glob.cmo                           \
-         lexers.cmo                         \
-         my_unix.cmo                        \
-         tags.cmo                           \
-         display.cmo                        \
-         log.cmo                            \
-         param_tags.cmo                     \
-         shell.cmo                          \
-         slurp.cmo                          \
-         ocamlbuild_where.cmo               \
-         command.cmo                        \
-         options.cmo  			    \
-         pathname.cmo                       \
-         digest_cache.cmo                   \
-         resource.cmo 			    \
-         rule.cmo 			    \
-         flags.cmo 			    \
-         solver.cmo 			    \
-         report.cmo 			    \
-         ocaml_arch.cmo 		    \
-         hygiene.cmo 			    \
-         configuration.cmo 		    \
-         tools.cmo 			    \
-         fda.cmo 			    \
-         plugin.cmo 			    \
-         ocaml_utils.cmo 		    \
-         ocaml_dependencies.cmo 	    \
-         ocaml_compiler.cmo 		    \
-         ocaml_tools.cmo 		    \
-         hooks.cmo 			    \
-         findlib.cmo 			    \
-         ocaml_specific.cmo 		    \
-         exit_codes.cmo 		    \
-         main.cmo
-
-BASE_CMOFILES= ocamlbuild_executor.cmo \
-	       ocamlbuild_unix_plugin.cmo
-
-INSTALL_LIBFILES = $(BASE_CMOFILES)           \
-		   $(BASE_CMOFILES:.cmo=.cmi) \
-		   $(OCAMLBUILD_LIBCMA)       \
-	           $(OCAMLBUILD).cmo          \
-		   $(OCAMLBUILD)_pack.cmi
-
-INSTALL_BINFILES = $(OCAMLBUILD)
-
-CMXFILES= $(CMOFILES:.cmo=.cmx)
-
-CMXFILES_PACK= $(CMOFILES_PACK:.cmo=.cmx)
-CMIFILES_PACK= $(CMOFILES_PACK:.cmo=.cmi) signatures.cmi
-
-EXECMOFILES_PACK= $(CMOFILES_PACK)
-EXECMXFILES_PACK= $(EXECMOFILES_PACK:.cmo=.cmx)
-EXECMIFILES_PACK= $(EXECMOFILES_PACK:.cmo=.cmi)
-
-LIBCMOFILES_PACK= $(CMOFILES_PACK)
-LIBCMXFILES_PACK= $(LIBCMOFILES_PACK:.cmo=.cmx)
-LIBCMIFILES_PACK= $(LIBCMOFILES_PACK:.cmo=.cmi)
-
-# Les cmo et cmx de la distrib OCAML
-OCAMLCMOFILES=
-OCAMLCMXFILES=$(OCAMLCMOFILES_PACK:.cmo=.cmx)
-
-all: exe lib
-opt: $(OCAMLBUILD).native
-exe: $(OCAMLBUILD)
-lib: $(OCAMLBUILD_LIBCMA)
-
-opt.opt: exeopt libopt
-exeopt: $(OCAMLBUILD_OPT)
-libopt: $(OCAMLBUILD_LIBCMXA) $(OCAMLBUILD_LIBCMI)
-
-debug:
-	$(MAKE) OCAMLPP=""
-
-$(OCAMLBUILD)_pack.cmo: $(CMOFILES_PACK) $(CMIFILES_PACK)
-	$(OCAMLC) -pack -o $@ $(LINKFLAGS) $(OCAMLCMOFILES_PACK) $(EXECMOFILES_PACK) signatures.mli
-
-$(OCAMLBUILD)_pack.cmx: $(EXECMXFILES_PACK)
-	$(OCAMLOPT) -pack -o $@ $(LINKFLAGS) $(OCAMLCMOFILES_PACK) $(EXECMXFILES_PACK)
-
-$(OCAMLBUILD): $(OCAMLBUILD)_pack.cmo $(CMOFILES) $(OCAMLBUILD).cmo $(BASE_CMOFILES)
-	$(OCAMLC) -o $@ unix.cma $(LINKFLAGS) $(OCAMLBUILD)_pack.cmo $(CMOFILES)
-
-$(OCAMLBUILD).native: $(OCAMLBUILD)_pack.cmx $(CMXFILES)
-	$(OCAMLOPT) -o $@  $(LINKFLAGS) $(CMXFILES)
-
-$(OCAMLBUILD_LIBCMA): $(LIBCMOFILES_PACK)
-	$(OCAMLC) -a -o $@ $(LINKFLAGS) $(OCAMLSRCDIR)/tools/depend.cmo $(LIBCMOFILES_PACK)
-$(OCAMLBUILD_LIBCMXA): $(LIBCMXFILES)
-	$(OCAMLOPT) -a -o $@ $(LINKFLAGS)	$(OCAMLSRCDIR)/tools/depend.cmx $(LIBCMXFILES)
-
-# generic rules :
-#################
-
-.SUFFIXES: .mll .mly .ml .mli .cmo .cmi .cmx .cmxs
-
-.ml.cmo:
-	$(OCAMLC) $(OCAMLPP) $(COMPFLAGS) -c $<
-
-.mli.cmi:
-	$(OCAMLC) $(OCAMLPP) $(COMPFLAGS) -c $<
-
-.ml.cmx:
-	$(OCAMLOPT) $(OCAMLPP) $(COMPFLAGS) -c $<
-
-.ml.cmxs:
-	$(OCAMLOPT) -shared -o $@ $(OCAMLPP) $(COMPFLAGS) $<
-
-.mll.ml:
-	$(OCAMLLEX) $<
-
-.mly.ml:
-	$(OCAMLYACC) -v $<
-
-.mly.mli:
-	$(OCAMLYACC) -v $<
-
-# Installation targets
-######################
-install: dummy
-	if test -d $(INSTALL_BINDIR); then : ; else $(MKDIR) $(INSTALL_BINDIR); fi
-	if test -d $(INSTALL_LIBDIR); then : ; else $(MKDIR) $(INSTALL_LIBDIR); fi
-	if test -d $(INSTALL_CUSTOMDIR); then : ; else $(MKDIR) $(INSTALL_CUSTOMDIR); fi
-	$(CP) $(OCAMLBUILD) $(INSTALL_BINDIR)/$(OCAMLBUILD)$(EXE)
-	$(CP) $(INSTALL_LIBFILES) $(INSTALL_LIBDIR)
-	$(CP) $(INSTALL_BINFILES) $(INSTALL_BINDIR)
-
-installopt:
-	if test -f $(OCAMLBUILD_OPT) ; then $(MAKE) installopt_really ; fi
-
-installopt_really:
-	if test -d $(INSTALL_BINDIR); then : ; else $(MKDIR) $(INSTALL_BINDIR); fi
-	if test -d $(INSTALL_LIBDIR); then : ; else $(MKDIR) $(INSTALL_LIBDIR); fi
-	$(CP) ocamlbuild.hva $(OCAMLBUILD_LIBA) $(OCAMLBUILD_LIBCMXA) $(INSTALL_LIBDIR)
-	$(CP) $(INSTALL_MLIS) $(INSTALL_CMIS) $(INSTALL_LIBDIR)
-
-
-# backup, clean and depend :
-############################
-
-clean:: dummy
-	@rm -f *~ \#*\#
-	@rm -f $(OCAMLBUILD) $(OCAMLBUILD_OPT) *.cma *.cmxa *.cmo *.cmi *.cmx *.$(A) *.$(O)
-	@rm -f glob_lexer.ml lexers.ml
-
-depend::
-	$(OCAMLDEP) $(INCLUDES_DEP) *.mli *.mll *.mly *.ml > .depend
-
-dummy:
-
-include .depend
-
-# Additional rules
-glob_lexer.cmo: glob_lexer.cmi
-lexers.cmo: lexers.cmi
-
-glob_lexer.cmx: glob_lexer.cmi
-lexers.cmx: lexers.cmi
diff --git a/ocamldoc/Makefile b/ocamldoc/Makefile
index 7a487c6..a5fa7f1 100644
--- a/ocamldoc/Makefile
+++ b/ocamldoc/Makefile
@@ -15,12 +15,10 @@ include ../config/Makefile
 # Various commands and dir
 ##########################
 ROOTDIR   = ..
-OCAMLRUN  = $(ROOTDIR)/boot/ocamlrun
-OCAMLC    = $(OCAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLOPT  = $(OCAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLDEP  = $(OCAMLRUN) $(ROOTDIR)/tools/ocamldep
-OCAMLLEX  = $(OCAMLRUN) $(ROOTDIR)/boot/ocamllex
-OCAMLYACC = $(ROOTDIR)/yacc/ocamlyacc
+OCAMLC    = $(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLOPT  = $(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLDEP  = $(CAMLRUN) $(ROOTDIR)/tools/ocamldep
+OCAMLLEX  = $(CAMLRUN) $(ROOTDIR)/boot/ocamllex
 OCAMLLIB  = $(LIBDIR)
 OCAMLBIN  = $(BINDIR)
 
@@ -173,7 +171,7 @@ debug:
 	$(MAKE) OCAMLPP=""
 
 $(OCAMLDOC): $(EXECMOFILES)
-	$(OCAMLC) -o $@ -linkall unix.cma str.cma dynlink.cma \
+	$(OCAMLC) -use-runtime $(CAMLRUN) -o $@ -linkall unix.cma str.cma dynlink.cma \
 	          $(OCAMLSRCDIR)/compilerlibs/ocamlcommon.cma \
 	          $(LINKFLAGS) $(OCAMLCMOFILES) $(EXECMOFILES)
 $(OCAMLDOC_OPT): $(EXECMXFILES)
@@ -233,10 +231,10 @@ odoc_see_lexer.ml: odoc_see_lexer.mll
 	$(OCAMLLEX) $<
 
 .mly.ml:
-	$(OCAMLYACC) -v $<
+	$(CAMLYACC) -v $<
 
 .mly.mli:
-	$(OCAMLYACC) -v $<
+	$(CAMLYACC) -v $<
 
 # Installation targets
 ######################
@@ -343,8 +341,8 @@ clean:: dummy
 	@rm -f generators/*.cm[aiox] generators/*.$(A) generators/*.$(O) generators/*.cmx[as]
 
 depend::
-	$(OCAMLYACC) odoc_text_parser.mly
-	$(OCAMLYACC) odoc_parser.mly
+	$(CAMLYACC) odoc_text_parser.mly
+	$(CAMLYACC) odoc_parser.mly
 	$(OCAMLLEX) odoc_text_lexer.mll
 	$(OCAMLLEX) odoc_lexer.mll
 	$(OCAMLLEX) odoc_ocamlhtml.mll
diff --git a/ocamldoc/Makefile.nt b/ocamldoc/Makefile.nt
index 22cd36e..6f6744d 100644
--- a/ocamldoc/Makefile.nt
+++ b/ocamldoc/Makefile.nt
@@ -15,12 +15,10 @@ include ../config/Makefile
 # Various commands and dir
 ##########################
 ROOTDIR   = ..
-OCAMLRUN  = $(ROOTDIR)/boot/ocamlrun
-OCAMLC    = $(OCAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLOPT  = $(OCAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLDEP  = $(OCAMLRUN) $(ROOTDIR)/tools/ocamldep
-OCAMLLEX  = $(OCAMLRUN) $(ROOTDIR)/boot/ocamllex
-OCAMLYACC = $(ROOTDIR)/yacc/ocamlyacc
+OCAMLC    = $(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLOPT  = $(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLDEP  = $(CAMLRUN) $(ROOTDIR)/tools/ocamldep
+OCAMLLEX  = $(CAMLRUN) $(ROOTDIR)/boot/ocamllex
 OCAMLLIB  = $(LIBDIR)
 OCAMLBIN  = $(BINDIR)
 
@@ -202,10 +200,10 @@ odoc_see_lexer.ml: odoc_see_lexer.mll
 	$(OCAMLLEX) $<
 
 .mly.ml:
-	$(OCAMLYACC) -v $<
+	$(CAMLYACC) -v $<
 
 .mly.mli:
-	$(OCAMLYACC) -v $<
+	$(CAMLYACC) -v $<
 
 # Installation targets
 ######################
@@ -240,8 +238,8 @@ clean:: dummy
 	@rm -f generators/*.cm[aiox] generators/*.$(A) generators/*.$(O) generators/*.cmx[as]
 
 depend::
-	$(OCAMLYACC) odoc_text_parser.mly
-	$(OCAMLYACC) odoc_parser.mly
+	$(CAMLYACC) odoc_text_parser.mly
+	$(CAMLYACC) odoc_parser.mly
 	$(OCAMLLEX) odoc_text_lexer.mll
 	$(OCAMLLEX) odoc_lexer.mll
 	$(OCAMLLEX) odoc_ocamlhtml.mll
diff --git a/otherlibs/Makefile b/otherlibs/Makefile
index 397497d..c18dc8c 100644
--- a/otherlibs/Makefile
+++ b/otherlibs/Makefile
@@ -13,8 +13,8 @@
 
 # Common Makefile for otherlibs on the Unix ports
 
-CAMLC=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
-CAMLOPT=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/ocamlopt -nostdlib \
+CAMLC=$(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
+CAMLOPT=$(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib \
         -I $(ROOTDIR)/stdlib
 CFLAGS=-I$(ROOTDIR)/byterun -O $(SHAREDCCCOMPOPTS) $(EXTRACFLAGS)
 
diff --git a/otherlibs/Makefile.shared b/otherlibs/Makefile.shared
index 9bed5f7..a76a092 100644
--- a/otherlibs/Makefile.shared
+++ b/otherlibs/Makefile.shared
@@ -18,7 +18,6 @@ include $(ROOTDIR)/config/Makefile
 
 # Compilation options
 CC=$(BYTECC)
-CAMLRUN=$(ROOTDIR)/boot/ocamlrun
 COMPFLAGS=-w +33..39 -warn-error A -bin-annot -g -safe-string $(EXTRACAMLFLAGS)
 MKLIB=$(CAMLRUN) $(ROOTDIR)/tools/ocamlmklib
 
diff --git a/otherlibs/bigarray/Makefile b/otherlibs/bigarray/Makefile
index 3f9afd5..3bcc7a4 100644
--- a/otherlibs/bigarray/Makefile
+++ b/otherlibs/bigarray/Makefile
@@ -22,6 +22,6 @@ include ../Makefile
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/bigarray/Makefile.nt b/otherlibs/bigarray/Makefile.nt
index 32642e1..baeaa7a 100644
--- a/otherlibs/bigarray/Makefile.nt
+++ b/otherlibs/bigarray/Makefile.nt
@@ -22,6 +22,6 @@ include ../Makefile.nt
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/dynlink/Makefile b/otherlibs/dynlink/Makefile
index 6284a52..2411369 100644
--- a/otherlibs/dynlink/Makefile
+++ b/otherlibs/dynlink/Makefile
@@ -16,9 +16,8 @@
 include ../../config/Makefile
 
 ROOTDIR   = ../..
-OCAMLRUN  = $(ROOTDIR)/boot/ocamlrun
-OCAMLC    = $(OCAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
-OCAMLOPT  = $(OCAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLC    = $(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib -I $(ROOTDIR)/stdlib
+OCAMLOPT  = $(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib -I $(ROOTDIR)/stdlib
 
 INCLUDES=-I ../../utils -I ../../typing -I ../../bytecomp -I ../../asmcomp
 COMPFLAGS=-w +33..39 -warn-error A -bin-annot -safe-string \
@@ -69,7 +68,7 @@ dynlink.cmx: dynlink.cmi natdynlink.ml
 	rm -f dynlink.mlopt
 
 extract_crc: dynlink.cma extract_crc.cmo
-	$(OCAMLC) $(COMPFLAGS) -o extract_crc dynlink.cma extract_crc.cmo
+	$(OCAMLC) -use-runtime $(CAMLRUN) -o extract_crc dynlink.cma extract_crc.cmo
 
 INSTALL_LIBDIR=$(DESTDIR)$(LIBDIR)
 
diff --git a/otherlibs/graph/Makefile b/otherlibs/graph/Makefile
index 2fb4580..850e025 100644
--- a/otherlibs/graph/Makefile
+++ b/otherlibs/graph/Makefile
@@ -27,6 +27,6 @@ include ../Makefile
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c | sed -e 's, /[^ ]*\.h,,g' > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/num/Makefile b/otherlibs/num/Makefile
index c867b2c..e08e029 100644
--- a/otherlibs/num/Makefile
+++ b/otherlibs/num/Makefile
@@ -32,6 +32,6 @@ bng.$(O): bng.h bng_digit.c \
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/str/Makefile b/otherlibs/str/Makefile
index 4a58d46..93b2bf9 100644
--- a/otherlibs/str/Makefile
+++ b/otherlibs/str/Makefile
@@ -28,6 +28,6 @@ str.cmx: str.cmi
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/systhreads/Makefile b/otherlibs/systhreads/Makefile
index c5b4e10..e9a46ef 100644
--- a/otherlibs/systhreads/Makefile
+++ b/otherlibs/systhreads/Makefile
@@ -14,11 +14,11 @@
 include ../../config/Makefile
 
 ROOTDIR=../..
-CAMLC=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/ocamlc -nostdlib \
+CAMLC=$(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib \
       -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix
-CAMLOPT=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/ocamlopt -nostdlib \
+CAMLOPT=$(CAMLRUN) $(ROOTDIR)/ocamlopt -nostdlib \
         -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix
-MKLIB=../../boot/ocamlrun ../../tools/ocamlmklib
+MKLIB=$(CAMLRUN) ../../tools/ocamlmklib
 COMPFLAGS=-w +33..39 -warn-error A -g -bin-annot -safe-string
 
 BYTECODE_C_OBJS=st_stubs_b.o
@@ -107,6 +107,6 @@ installopt:
 
 depend: $(GENFILES)
 	-$(CC) -MM -I../../byterun *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/systhreads/Makefile.nt b/otherlibs/systhreads/Makefile.nt
index 3411761..9d74390 100644
--- a/otherlibs/systhreads/Makefile.nt
+++ b/otherlibs/systhreads/Makefile.nt
@@ -14,10 +14,10 @@
 include ../../config/Makefile
 
 # Compilation options
-CAMLC=../../boot/ocamlrun ../../ocamlc -I ../../stdlib -I ../win32unix
-CAMLOPT=../../boot/ocamlrun ../../ocamlopt -I ../../stdlib -I ../win32unix
+CAMLC=$(CAMLRUN) ../../ocamlc -I ../../stdlib -I ../win32unix
+CAMLOPT=$(CAMLRUN) ../../ocamlopt -I ../../stdlib -I ../win32unix
 COMPFLAGS=-w +33 -warn-error A -g
-MKLIB=../../boot/ocamlrun ../../tools/ocamlmklib
+MKLIB=$(CAMLRUN) ../../tools/ocamlmklib
 CFLAGS=-I../../byterun $(EXTRACFLAGS)
 
 CAMLOBJS=thread.cmo mutex.cmo condition.cmo event.cmo threadUnix.cmo
@@ -32,7 +32,7 @@ all: lib$(LIBNAME).$(A) $(LIBNAME).cma $(CMIFILES)
 allopt: lib$(LIBNAME).$(A) $(LIBNAME).cmxa $(LIBNAME).cmxs $(CMIFILES)
 
 $(LIBNAME).cma: $(CAMLOBJS)
-	$(MKLIB) -o $(LIBNAME) -ocamlc "../../boot/ocamlrun ../../ocamlc" \
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLRUN) ../../ocamlc" \
 	         -linkall $(CAMLOBJS) $(LINKOPTS)
 
 lib$(LIBNAME).$(A): $(COBJS)
@@ -46,7 +46,7 @@ st_stubs_b.$(O): st_stubs.c st_win32.h
 
 $(LIBNAME).cmxa: $(CAMLOBJS:.cmo=.cmx)
 	$(MKLIB) -o $(LIBNAME)nat \
-	         -ocamlopt "../../boot/ocamlrun ../../ocamlopt" -linkall \
+	         -ocamlopt "$(CAMLRUN) ../../ocamlopt" -linkall \
 	         $(CAMLOBJS:.cmo=.cmx) $(LINKOPTS)
 	mv $(LIBNAME)nat.cmxa $(LIBNAME).cmxa
 	mv $(LIBNAME)nat.$(A) $(LIBNAME).$(A)
diff --git a/otherlibs/threads/Makefile b/otherlibs/threads/Makefile
index d141815..869def4 100644
--- a/otherlibs/threads/Makefile
+++ b/otherlibs/threads/Makefile
@@ -122,6 +122,6 @@ installopt:
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/otherlibs/unix/Makefile b/otherlibs/unix/Makefile
index e41a1e1..faebd3f 100644
--- a/otherlibs/unix/Makefile
+++ b/otherlibs/unix/Makefile
@@ -42,6 +42,6 @@ include ../Makefile
 
 depend:
 	$(CC) -MM $(CFLAGS) *.c > .depend
-	../../boot/ocamlrun ../../tools/ocamldep *.mli *.ml >> .depend
+	$(CAMLRUN) ../../tools/ocamldep *.mli *.ml >> .depend
 
 include .depend
diff --git a/stdlib/Makefile.shared b/stdlib/Makefile.shared
index 5f5761f..ac9c43c 100755
--- a/stdlib/Makefile.shared
+++ b/stdlib/Makefile.shared
@@ -14,11 +14,11 @@
 include ../config/Makefile
 RUNTIME=../boot/ocamlrun
 COMPILER=../ocamlc
-CAMLC=$(RUNTIME) $(COMPILER)
+CAMLC=$(CAMLRUN) $(COMPILER)
 COMPFLAGS=-strict-sequence -w +33..39 -g -warn-error A -bin-annot -nostdlib \
           -safe-string
 OPTCOMPILER=../ocamlopt
-CAMLOPT=$(RUNTIME) $(OPTCOMPILER)
+CAMLOPT=$(CAMLRUN) $(OPTCOMPILER)
 CAMLDEP=../boot/ocamlrun ../tools/ocamldep
 
 OBJS=camlinternalFormatBasics.cmo pervasives.cmo $(OTHERS)
diff --git a/tools/Makefile.nt b/tools/Makefile.nt
index ed9b359..13636d0 100644
--- a/tools/Makefile.nt
+++ b/tools/Makefile.nt
@@ -18,7 +18,7 @@ OCAMLMKTOP=ocamlmktop.cmo
 OCAMLMKTOP_IMPORTS=misc.cmo config.cmo clflags.cmo ccomp.cmo
 
 ocamlmktop: $(OCAMLMKTOP)
-	$(CAMLC) $(LINKFLAGS) -o ocamlmktop $(OCAMLMKTOP_IMPORTS) $(OCAMLMKTOP)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o ocamlmktop $(OCAMLMKTOP_IMPORTS) $(OCAMLMKTOP)
 
 install::
 	cp ocamlmktop $(INSTALL_BINDIR)/ocamlmktop$(EXE)
diff --git a/tools/Makefile.shared b/tools/Makefile.shared
index 302e147..8d2ffed 100644
--- a/tools/Makefile.shared
+++ b/tools/Makefile.shared
@@ -12,7 +12,6 @@
 
 include ../config/Makefile
 
-CAMLRUN=../boot/ocamlrun
 CAMLC=$(CAMLRUN) ../boot/ocamlc -nostdlib -I ../boot
 CAMLOPT=$(CAMLRUN) ../ocamlopt -nostdlib -I ../stdlib
 CAMLLEX=$(CAMLRUN) ../boot/ocamllex
@@ -40,7 +39,8 @@ CAMLDEP_IMPORTS=misc.cmo config.cmo clflags.cmo terminfo.cmo \
   ccomp.cmo ast_mapper.cmo pparse.cmo compenv.cmo
 
 ocamldep: depend.cmi $(CAMLDEP_OBJ)
-	$(CAMLC) $(LINKFLAGS) -compat-32 -o ocamldep $(CAMLDEP_IMPORTS) $(CAMLDEP_OBJ)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -compat-32 \
+           -o ocamldep $(CAMLDEP_IMPORTS) $(CAMLDEP_OBJ)
 
 ocamldep.opt: depend.cmi $(CAMLDEP_OBJ:.cmo=.cmx)
 	$(CAMLOPT) $(LINKFLAGS) -o ocamldep.opt $(CAMLDEP_IMPORTS:.cmo=.cmx) \
@@ -69,14 +69,16 @@ CSLPROF_IMPORTS=misc.cmo config.cmo clflags.cmo terminfo.cmo \
   syntaxerr.cmo ast_helper.cmo parser.cmo lexer.cmo parse.cmo
 
 ocamlprof: $(CSLPROF) profiling.cmo
-	$(CAMLC) $(LINKFLAGS) -o ocamlprof $(CSLPROF_IMPORTS) $(CSLPROF)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o ocamlprof $(CSLPROF_IMPORTS) $(CSLPROF)
 
 ocamlcp: ocamlcp.cmo
-	$(CAMLC) $(LINKFLAGS) -o ocamlcp warnings.cmo main_args.cmo ocamlcp.cmo
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o ocamlcp warnings.cmo main_args.cmo ocamlcp.cmo
 
 ocamloptp: ocamloptp.cmo
-	$(CAMLC) $(LINKFLAGS) -o ocamloptp warnings.cmo main_args.cmo \
-	         ocamloptp.cmo
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o ocamloptp warnings.cmo main_args.cmo ocamloptp.cmo
 
 opt:: profiling.cmx
 
@@ -96,7 +98,8 @@ clean::
 # To help building mixed-mode libraries (OCaml + C)
 
 ocamlmklib: ocamlmklibconfig.cmo ocamlmklib.cmo
-	$(CAMLC) $(LINKFLAGS) -o ocamlmklib ocamlmklibconfig.cmo ocamlmklib.cmo
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o ocamlmklib ocamlmklibconfig.cmo ocamlmklib.cmo
 
 install::
 	cp ocamlmklib $(INSTALL_BINDIR)/ocamlmklib$(EXE)
@@ -128,7 +131,8 @@ OCAML299TO3= lexer299.cmo ocaml299to3.cmo
 LIBRARY3= misc.cmo warnings.cmo location.cmo
 
 ocaml299to3: $(OCAML299TO3)
-	$(CAMLC) $(LINKFLAGS) -o ocaml299to3 $(LIBRARY3) $(OCAML299TO3)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o ocaml299to3 $(LIBRARY3) $(OCAML299TO3)
 
 lexer299.ml: lexer299.mll
 	$(CAMLLEX) lexer299.mll
@@ -144,7 +148,8 @@ clean::
 SCRAPELABELS= lexer301.cmo scrapelabels.cmo
 
 scrapelabels: $(SCRAPELABELS)
-	$(CAMLC) $(LINKFLAGS) -o scrapelabels $(LIBRARY3) $(SCRAPELABELS)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) \
+           -o scrapelabels $(LIBRARY3) $(SCRAPELABELS)
 
 lexer301.ml: lexer301.mll
 	$(CAMLLEX) lexer301.mll
@@ -162,8 +167,8 @@ ADDLABELS_IMPORTS=misc.cmo config.cmo clflags.cmo terminfo.cmo \
   syntaxerr.cmo ast_helper.cmo parser.cmo lexer.cmo parse.cmo
 
 addlabels: addlabels.cmo
-	$(CAMLC) $(LINKFLAGS) -w sl -o addlabels \
-		$(ADDLABELS_IMPORTS) addlabels.cmo
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -w sl \
+           -o addlabels $(ADDLABELS_IMPORTS) addlabels.cmo
 
 #install::
 #	cp addlabels $(INSTALL_LIBDIR)
@@ -176,7 +181,7 @@ clean::
 CVT_EMIT=cvt_emit.cmo
 
 cvt_emit: $(CVT_EMIT)
-	$(CAMLC) $(LINKFLAGS) -o cvt_emit $(CVT_EMIT)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o cvt_emit $(CVT_EMIT)
 
 # cvt_emit is precious: sometimes we are stuck in the middle of a
 # bootstrap and we need to remake the dependencies
@@ -231,7 +236,7 @@ READ_CMT= \
           cmt2annot.cmo read_cmt.cmo
 
 read_cmt: $(READ_CMT)
-	$(CAMLC) $(LINKFLAGS) -o read_cmt $(READ_CMT)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o read_cmt $(READ_CMT)
 
 read_cmt.opt: $(READ_CMT:.cmo=.cmx)
 	$(CAMLOPT) $(LINKFLAGS) -o read_cmt.opt $(READ_CMT:.cmo=.cmx)
@@ -246,7 +251,7 @@ beforedepend::
 DUMPOBJ=opnames.cmo dumpobj.cmo
 
 dumpobj: $(DUMPOBJ)
-	$(CAMLC) $(LINKFLAGS) -o dumpobj \
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o dumpobj \
 	         misc.cmo tbl.cmo config.cmo ident.cmo \
 	         opcodes.cmo bytesections.cmo $(DUMPOBJ)
 
@@ -282,7 +287,7 @@ OBJINFO=../compilerlibs/ocamlcommon.cma \
         objinfo.cmo
 
 objinfo: objinfo_helper$(EXE) $(OBJINFO)
-	$(CAMLC) -o objinfo $(OBJINFO)
+	$(CAMLC) -use-runtime $(CAMLRUN) -o objinfo $(OBJINFO)
 
 install::
 	cp objinfo $(INSTALL_BINDIR)/ocamlobjinfo$(EXE)
@@ -296,7 +301,7 @@ clean::
 PRIMREQ=primreq.cmo
 
 primreq: $(PRIMREQ)
-	$(CAMLC) $(LINKFLAGS) -o primreq config.cmo $(PRIMREQ)
+	$(CAMLC) $(LINKFLAGS) -use-runtime $(CAMLRUN) -o primreq config.cmo $(PRIMREQ)
 
 clean::
 	rm -f primreq
