--- Makefile.orig	2016-06-10 12:37:48.000000000 +0000
+++ Makefile	2016-11-24 08:14:29.609048303 +0000
@@ -25,9 +25,15 @@
 OBJS=zlib.cmo zip.cmo gzip.cmo
 C_OBJS=zlibstubs.o

+NATDYNLINK := $(shell if [ -f `ocamlc -where`/dynlink.cmxa ]; then echo YES; else echo NO; fi)
+
+ifeq "${NATDYNLINK}" "YES"
+CMXS = zip.cmxs
+endif
+
 all: libcamlzip.a zip.cma

-allopt: libcamlzip.a zip.cmxa
+allopt: libcamlzip.a zip.cmxa $(CMXS)

 zip.cma: $(OBJS)
	$(OCAMLMKLIB) -o zip -oc camlzip $(OBJS) \
@@ -37,6 +43,9 @@
	$(OCAMLMKLIB) -o zip -oc camlzip $(OBJS:.cmo=.cmx) \
             -L$(ZLIB_LIBDIR) $(ZLIB_LIB)

+zip.cmxs: zip.cmxa
+	$(OCAMLOPT) -shared -linkall -I ./ -o $@ $^
+
 libcamlzip.a: $(C_OBJS)
	$(OCAMLMKLIB) -oc camlzip $(C_OBJS) \
             -L$(ZLIB_LIBDIR) $(ZLIB_LIB)
@@ -58,7 +67,7 @@

 install:
	mkdir -p $(INSTALLDIR)
-	cp zip.cma zip.cmi gzip.cmi zip.mli gzip.mli libcamlzip.a $(INSTALLDIR)
+	cp zip.cma zip.cmi gzip.cmi zlib.cmi zip.mli gzip.mli zlib.mli libcamlzip.a $(INSTALLDIR)
	if test -f dllcamlzip.so; then \
	  cp dllcamlzip.so $(INSTALLDIR); \
           ldconf=`$(OCAMLC) -where`/ld.conf; \
@@ -68,10 +77,11 @@
         fi

 installopt:
-	cp zip.cmxa zip.a zip.cmx gzip.cmx $(INSTALLDIR)
+	cp zip.cmxa $(CMXS) zip.a zip.cmx gzip.cmx zlib.cmx $(INSTALLDIR)

 install-findlib:
-	ocamlfind install zip META *.mli *.a *.cmi *.cma $(wildcard *.cmxa) $(wildcard *.so)
+	ocamlfind install zip META *.mli *.a *.cmi *.cma $(wildcard *.cmx) $(wildcard *.cmxa) $(wildcard *.cmxs) $(wildcard *.so)
+	echo 'directory="../zip"' >> META && ocamlfind install camlzip META

 depend:
	gcc -MM -I$(ZLIB_INCLUDE) *.c > .depend
