From 9dca70513fba848e1ecee65c7f921e0d5491bea0 Mon Sep 17 00:00:00 2001
From: Peter Zotov <whitequark@whitequark.org>
Date: Sat, 18 Oct 2014 11:52:02 +0400
Subject: [PATCH] Invoke ocamldep and ocamlmklib through ocamlfind.

---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 48af4c5..87c83aa 100644
--- a/Makefile
+++ b/Makefile
@@ -2,9 +2,9 @@
 
 DEBUG=false
 OCAML=ocaml
-OCAMLDEP=ocamldep
 OCAMLFIND=ocamlfind
-OCAMLMKLIB=ocamlmklib
+OCAMLDEP=$(OCAMLFIND) ocamldep
+OCAMLMKLIB=$(OCAMLFIND) ocamlmklib
 VPATH=src examples
 BUILDDIR=_build
 PROJECTS=configure libffi-abigen configured ctypes cstubs ctypes-foreign-base ctypes-foreign-threaded ctypes-foreign-unthreaded ctypes-top
-- 
2.1.1

