--- Makefile.orig	2016-03-19 01:52:32.051579250 +0000
+++ Makefile	2016-03-19 01:52:58.831625938 +0000
@@ -3,22 +3,16 @@
 # The name of the Zlib library.  Usually -lz
 ZLIB_LIB=-lz
 
-# The directory containing the Zlib library (libz.a or libz.so)
-ZLIB_LIBDIR=/usr/local/lib
-
-# The directory containing the Zlib header file (zlib.h)
-ZLIB_INCLUDE=/usr/local/include
-
 # Where to install the library.  By default: sub-directory 'zip' of
 # OCaml's standard library directory.
 INSTALLDIR=`$(OCAMLC) -where`/zip
 
 ### End of configuration section
 
-OCAMLC=ocamlc -g
-OCAMLOPT=ocamlopt
-OCAMLDEP=ocamldep
-OCAMLMKLIB=ocamlmklib
+OCAMLC=ocamlfind ocamlc -g
+OCAMLOPT=ocamlfind ocamlopt
+OCAMLDEP=ocamlfind ocamldep
+OCAMLMKLIB=ocamlfind ocamlmklib
 
 OBJS=zlib.cmo zip.cmo gzip.cmo
 C_OBJS=zlibstubs.o
@@ -35,18 +29,18 @@
 
 zip.cma: $(OBJS)
 	$(OCAMLMKLIB) -o zip -oc camlzip $(OBJS) \
-            -L$(ZLIB_LIBDIR) $(ZLIB_LIB)
+            $(ZLIB_LIB)
 
 zip.cmxa: $(OBJS:.cmo=.cmx)
 	$(OCAMLMKLIB) -o zip -oc camlzip $(OBJS:.cmo=.cmx) \
-            -L$(ZLIB_LIBDIR) $(ZLIB_LIB)
+            $(ZLIB_LIB)
 
 zip.cmxs: zip.cmxa
 	$(OCAMLOPT) -shared -linkall -I ./ -o $@ $^
 
 libcamlzip.a: $(C_OBJS)
 	$(OCAMLMKLIB) -oc camlzip $(C_OBJS) \
-            -L$(ZLIB_LIBDIR) $(ZLIB_LIB)
+            $(ZLIB_LIB)
 
 .SUFFIXES: .mli .ml .cmo .cmi .cmx
 
@@ -57,7 +51,7 @@
 .ml.cmx:
 	$(OCAMLOPT) -c $<
 .c.o:
-	$(OCAMLC) -c -ccopt -g -ccopt -I$(ZLIB_INCLUDE) $<
+	$(OCAMLC) -c $<
 
 clean:
 	rm -f *.cm*
@@ -78,11 +72,11 @@
 	cp zip.cmxa $(CMXS) zip.a zip.cmx gzip.cmx zlib.cmx $(INSTALLDIR)
 
 install-findlib:
 	ocamlfind install zip META *.mli *.a *.cmi *.cma $(wildcard *.cmx) $(wildcard *.cmxa) $(wildcard *.cmxs) $(wildcard *.so)
 	echo 'directory="../zip"' >> META && ocamlfind install camlzip META
 
 depend:
-	gcc -MM -I$(ZLIB_INCLUDE) *.c > .depend
+	gcc -MM *.c > .depend
 	ocamldep *.mli *.ml >> .depend
 
 include .depend
