--- a/lib/sqlite3_stubs.c   2014-12-26 18:32:41.457159414 +0300
+++ b/lib/sqlite3_stubs.c   2014-12-26 18:32:55.753151835 +0300
@@ -36,7 +36,7 @@
 #include <caml/custom.h>
 #include <caml/signals.h>
 
-#include <sqlite3.h>
+#include "sqlite3_db.h"
 
 #if __GNUC__ >= 3
 # define inline inline __attribute__ ((always_inline))
--- a/myocamlbuild.ml	2014-10-08 19:59:05.000000000 +0400
+++ b/myocamlbuild.ml	2014-12-26 18:44:51.968726282 +0300
@@ -638,10 +638,6 @@
                       A "-Wno-long-long"
                    ])
             ]);
-          (["oasis_library_sqlite3_cclib"; "link"],
-            [(OASISExpr.EBool true, S [A "-cclib"; A "-lsqlite3"])]);
-          (["oasis_library_sqlite3_cclib"; "ocamlmklib"; "c"],
-            [(OASISExpr.EBool true, S [A "-lsqlite3"])])
        ];
      includes = [("test", ["lib"])]
   }
@@ -694,44 +690,7 @@
 let () =
   let additional_rules = function
     | After_rules ->
-        (* Add correct SQLite3 compilation and link flags *)
-        let rec split_string s =
-          match try Some (String.index s ' ') with Not_found -> None with
-          | Some pos ->
-              let before = string_trim (String.before s pos) in
-              let after = String.after s (pos + 1) in
-              if before = "" then split_string after
-              else before :: split_string after
-          | None when s = "" -> []
-          | None -> [s]
-        in
-        let ocamlify ~ocaml_flag flags =
-          let chunks = split_string flags in
-          let cnv flag = [A ocaml_flag; A flag] in
-          List.concat (List.map cnv chunks)
-        in
-        let split_flags flags =
-          let chunks = split_string flags in
-          let cnv flag = A flag in
-          List.map cnv chunks
-        in
-        let osqlite3_cflags =
-          let cmd = "pkg-config --cflags sqlite3" in
-          match read_lines_from_cmd ~max_lines:1 cmd with
-          | [cflags] -> S (ocamlify ~ocaml_flag:"-ccopt" cflags)
-          | _ -> failwith "pkg-config failed for cflags"
-        in
-        let sqlite3_clibs, osqlite3_clibs =
-          let cmd = "pkg-config --libs sqlite3" in
-          match read_lines_from_cmd ~max_lines:1 cmd with
-          | [libs] ->
-              S (split_flags libs), S (ocamlify ~ocaml_flag:"-cclib" libs)
-          | _ -> failwith "pkg-config failed for libs"
-        in
-        flag ["compile"; "c"] osqlite3_cflags;
-        flag ["link"; "ocaml"; "library"] osqlite3_clibs;
-        flag ["oasis_library_sqlite3_cclib"; "ocamlmklib"; "c"] sqlite3_clibs;
-        flag ["oasis_library_sqlite3_cclib"; "link"] osqlite3_clibs
+        dep ["file:lib/sqlite3_stubs.c"] ["lib/sqlite3_db.h"]
       | _ -> ()
   in
   dispatch (
