From 531a142624e12468471ccc76a04c76ebaa6849ba Mon Sep 17 00:00:00 2001
From: Peter Zotov <whitequark@whitequark.org>
Date: Sat, 18 Oct 2014 11:41:17 +0400
Subject: [PATCH] Since r10c, Android NDK includes sys/ucontext.h.

---
 asmrun/signals_osdep.h | 13 +------------
 1 file changed, 1 insertion(+), 12 deletions(-)

diff --git a/asmrun/signals_osdep.h b/asmrun/signals_osdep.h
index f3b4642..8dd651f 100644
--- a/asmrun/signals_osdep.h
+++ b/asmrun/signals_osdep.h
@@ -66,18 +66,7 @@
 #elif defined(TARGET_arm) && (defined(SYS_linux_eabi) \
       || defined(SYS_linux_eabihf))
 
-  #if defined(__ANDROID__)
-    // The Android NDK does not have sys/ucontext.h yet.
-    typedef struct ucontext {
-      uint32_t uc_flags;
-      struct ucontext *uc_link;
-      stack_t uc_stack;
-      struct sigcontext uc_mcontext;
-      // Other fields omitted...
-    } ucontext_t;
-  #else
-    #include <sys/ucontext.h>
-  #endif
+  #include <sys/ucontext.h>
 
   #define DECLARE_SIGNAL_HANDLER(name) \
     static void name(int sig, siginfo_t * info, ucontext_t * context)
-- 
2.1.1

