diff --git a/asmcomp/asmlink.ml b/asmcomp/asmlink.ml
index bc62b01..85b551d 100644
--- a/asmcomp/asmlink.ml
+++ b/asmcomp/asmlink.ml
@@ -285,12 +285,13 @@ let link_shared ppf objfiles output_name =
 let call_linker file_list startup_file output_name =
   let main_dll = !Clflags.output_c_object
                  && Filename.check_suffix output_name Config.ext_dll
+  and main_obj_runtime = !Clflags.output_c_object && !Clflags.runtime_variant <> ""
   in
   let files = startup_file :: (List.rev file_list) in
   let files, c_lib =
-    if (not !Clflags.output_c_object) || main_dll then
+    if (not !Clflags.output_c_object) || main_dll || main_obj_runtime then
       files @ (List.rev !Clflags.ccobjs) @ runtime_lib (),
-      (if !Clflags.nopervasives then "" else Config.native_c_libraries)
+      (if !Clflags.nopervasives || main_obj_runtime then "" else Config.native_c_libraries)
     else
       files, ""
   in
diff --git a/utils/ccomp.ml b/utils/ccomp.ml
index bfe077a..20195c3 100644
--- a/utils/ccomp.ml
+++ b/utils/ccomp.ml
@@ -98,14 +98,22 @@ type link_mode =
   | MainDll
   | Partial
 
+let remove_Wl cclibs =
+  cclibs |> List.map (fun cclib ->
+    (* -Wl,-foo,bar -> -foo bar *)
+    if "-Wl," = String.sub cclib 0 4 then
+      String.map (function ',' -> ' ' | c -> c)
+                 (String.sub cclib 4 (String.length cclib - 4))
+    else cclib)
+
 let call_linker mode output_name files extra =
-  let files = quote_files files in
   let cmd =
     if mode = Partial then
-      Printf.sprintf "%s%s %s %s"
+      Printf.sprintf "%s%s %s %s %s"
         Config.native_pack_linker
         (Filename.quote output_name)
-        files
+        (quote_prefixed "-L" !Config.load_path)
+        (quote_files (remove_Wl files))
         extra
     else
       Printf.sprintf "%s -o %s %s %s %s %s %s %s"
@@ -121,7 +129,7 @@ let call_linker mode output_name files extra =
         ""  (*(Clflags.std_include_flag "-I")*)
         (quote_prefixed "-L" !Config.load_path)
         (String.concat " " (List.rev !Clflags.all_ccopts))
-        files
+        (quote_files files)
         extra
   in
   command cmd = 0
